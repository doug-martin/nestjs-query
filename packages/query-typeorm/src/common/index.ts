export * from './randomString';
export * from './typeorm';
