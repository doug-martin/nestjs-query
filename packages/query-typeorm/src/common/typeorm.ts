import { Class } from '@nestjs-query/core';
import { ColumnType, Connection, ConnectionOptions, getConnection } from 'typeorm';

interface QueryTypeormPropertyMetadata {
  metaType: 'property';
  type: ColumnType;
}

interface QueryTypeormRelationMetadata {
  metaType: 'relation';
  type: Class<unknown>;
}

export type QueryTypeormEntityMetadata<T = unknown> = Record<
  keyof T | string,
  QueryTypeormPropertyMetadata | QueryTypeormRelationMetadata
>;
export type QueryTypeormMetadata = Map<Class<unknown>, QueryTypeormEntityMetadata>;

export function buildQueryTypeormMetadata(connection: Connection): QueryTypeormMetadata {
  const meta: QueryTypeormMetadata = new Map();
  for (const entity of connection.entityMetadatas) {
    const entityMeta: QueryTypeormEntityMetadata = {};
    for (const field of [...entity.ownColumns]) {
      entityMeta[field.propertyName] = {
        metaType: 'property',
        type: field.type,
      };
    }
    for (const field of [...entity.ownRelations]) {
      // Skip strings
      if (typeof field.inverseEntityMetadata.target === 'function') {
        entityMeta[field.propertyName] = {
          metaType: 'relation',
          type: field.inverseEntityMetadata.target as Class<unknown>,
        };
      }
    }

    // Ignore things like junction tables
    if (typeof entity.target === 'string') {
      continue;
    }
    meta.set(entity.target as Class<unknown>, entityMeta);
  }
  return meta;
}

function getConnectionFromOpts(connection?: Connection | ConnectionOptions | string): Connection {
  if (!connection) {
    return getConnection('default');
  }
  if (typeof connection === 'string') {
    return getConnection(connection);
  }
  if (!connection.name) {
    return getConnection('default');
  }
  return connection as Connection;
}

const cache = new Map<Connection, QueryTypeormMetadata>();

export function getQueryTypeormMetadata(
  connectionOpts?: Connection | ConnectionOptions | string,
): QueryTypeormMetadata {
  const connection = getConnectionFromOpts(connectionOpts);

  let meta = cache.get(connection);
  if (!meta) {
    meta = buildQueryTypeormMetadata(connection);
    cache.set(connection, meta);
  }
  return meta;
}
