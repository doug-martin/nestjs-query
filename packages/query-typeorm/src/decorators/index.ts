export { InjectTypeOrmQueryService } from './inject-typeorm-query-service.decorator';
