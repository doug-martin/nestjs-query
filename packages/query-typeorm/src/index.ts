export { TypeOrmQueryService, TypeOrmQueryServiceOpts } from './services';
export { NestjsQueryTypeOrmModule } from './module';
