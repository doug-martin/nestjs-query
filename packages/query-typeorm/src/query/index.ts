export * from './filter-query.builder';
export * from './where.builder';
export * from './sql-comparison.builder';
export * from './relation-query.builder';
export * from './aggregate.builder';
