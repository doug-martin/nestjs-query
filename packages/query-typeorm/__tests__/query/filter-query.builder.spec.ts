import { anything, instance, mock, verify, when, deepEqual } from 'ts-mockito';
import { QueryBuilder, WhereExpression } from 'typeorm';
import { Class, Filter, Query, SortDirection, SortNulls } from '@nestjs-query/core';
import { closeTestConnection, createTestConnection, getTestConnection } from '../__fixtures__/connection.fixture';
import { TestSoftDeleteEntity } from '../__fixtures__/test-soft-delete.entity';
import { TestEntity } from '../__fixtures__/test.entity';
import { FilterQueryBuilder, WhereBuilder } from '../../src/query';

describe('FilterQueryBuilder', (): void => {
  beforeEach(createTestConnection);
  afterEach(closeTestConnection);

  const getEntityQueryBuilder = <Entity>(
    entity: Class<Entity>,
    whereBuilder: WhereBuilder<Entity>,
  ): FilterQueryBuilder<Entity> => new FilterQueryBuilder(getTestConnection().getRepository(entity), whereBuilder);

  const expectSQLSnapshot = <Entity>(query: QueryBuilder<Entity>): void => {
    const [sql, params] = query.getQueryAndParameters();
    expect(sql).toMatchSnapshot();
    expect(params).toMatchSnapshot();
  };

  describe('#getReferencedRelationsRecursive', () => {
    it('with deeply nested and / or', () => {
      const complexQuery: Filter<TestEntity> = {
        and: [
          {
            or: [
              { and: [{ stringType: { eq: '123' } }] },
              {
                and: [{ stringType: { eq: '123' } }, { id: { gt: '123' } }],
              },
            ],
          },
          {
            stringType: { eq: '345' },
            or: [
              { oneTestRelation: { relationName: { eq: '123' } } },
              { oneTestRelation: { relationOfTestRelation: { testRelationId: { eq: 'e1' } } } },
            ],
          },
        ],
      };
      const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
      const qb = getEntityQueryBuilder(TestEntity, instance(mockWhereBuilder));
      expect(qb.getReferencedRelationsRecursive(qb.repo.metadata, complexQuery)).toEqual({
        oneTestRelation: { relationOfTestRelation: {} },
      });
    });
    it('with nested and / or', () => {
      const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
      const qb = getEntityQueryBuilder(TestEntity, instance(mockWhereBuilder));
      expect(
        qb.getReferencedRelationsRecursive(qb.repo.metadata, {
          test: '123',
          and: [
            {
              boolType: { is: true },
            },
            {
              testRelations: {
                relationName: { eq: '123' },
              },
            },
          ],
          or: [
            {
              boolType: { is: true },
            },
            {
              oneTestRelation: {
                testRelationPk: { eq: '123' },
              },
            },
            {
              oneTestRelation: {
                relationsOfTestRelation: {
                  testRelationId: {
                    eq: '123',
                  },
                },
              },
            },
          ],
        } as Filter<TestEntity>),
      ).toEqual({ testRelations: {}, oneTestRelation: { relationsOfTestRelation: {} } });
    });
  });

  describe('#select', () => {
    const expectSelectSQLSnapshot = (query: Query<TestEntity>, whereBuilder: WhereBuilder<TestEntity>): void => {
      const selectQueryBuilder = getEntityQueryBuilder(TestEntity, whereBuilder).select(query);
      expectSQLSnapshot(selectQueryBuilder);
    };

    describe('with filter', () => {
      it('should not call whereBuilder#build', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot({}, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });

      it('should call whereBuilder#build if there is a filter', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        const query = { filter: { stringType: { eq: 'foo' } } };
        when(mockWhereBuilder.build(anything(), query.filter, deepEqual({}), 'TestEntity')).thenCall(
          (where: WhereExpression, field: Filter<TestEntity>, relationNames: string[], alias: string) =>
            where.andWhere(`${alias}.stringType = 'foo'`),
        );
        expectSelectSQLSnapshot(query, instance(mockWhereBuilder));
      });
    });

    describe('with paging', () => {
      it('should apply empty paging args', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot({}, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), deepEqual({}), 'TestEntity')).never();
      });

      it('should apply paging args going forward', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot({ paging: { limit: 10, offset: 11 } }, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), deepEqual({}), 'TestEntity')).never();
      });

      it('should apply paging args going backward', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot({ paging: { limit: 10, offset: 10 } }, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });
    });

    describe('with sorting', () => {
      it('should apply ASC sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.ASC }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });

      it('should apply ASC NULLS_FIRST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_FIRST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });

      it('should apply ASC NULLS_LAST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_LAST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });

      it('should apply DESC sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.DESC }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });

      it('should apply DESC NULLS_FIRST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_FIRST }] },
          instance(mockWhereBuilder),
        );
      });

      it('should apply DESC NULLS_LAST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_LAST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });

      it('should apply multiple sorts', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectSelectSQLSnapshot(
          {
            sorting: [
              { field: 'numberType', direction: SortDirection.ASC },
              { field: 'boolType', direction: SortDirection.DESC },
              { field: 'stringType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_FIRST },
              { field: 'dateType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_LAST },
            ],
          },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), {}, 'TestEntity')).never();
      });
    });
  });

  describe('#update', () => {
    const expectUpdateSQLSnapshot = (query: Query<TestEntity>, whereBuilder: WhereBuilder<TestEntity>): void => {
      const queryBuilder = getEntityQueryBuilder(TestEntity, whereBuilder).update(query).set({ stringType: 'baz' });
      expectSQLSnapshot(queryBuilder);
    };

    describe('with filter', () => {
      it('should call whereBuilder#build if there is a filter', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        const query = { filter: { stringType: { eq: 'foo' } } };
        when(mockWhereBuilder.build(anything(), query.filter, deepEqual({}), undefined)).thenCall(
          (where: WhereExpression) => where.andWhere(`stringType = 'foo'`),
        );
        expectUpdateSQLSnapshot(query, instance(mockWhereBuilder));
      });
    });
    describe('with paging', () => {
      it('should ignore paging args', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot({ paging: { limit: 10, offset: 11 } }, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });
    });

    describe('with sorting', () => {
      it('should apply ASC sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.ASC }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });

      it('should apply ASC NULLS_FIRST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_FIRST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });

      it('should apply ASC NULLS_LAST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_LAST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });

      it('should apply DESC sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.DESC }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });

      it('should apply DESC NULLS_FIRST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_FIRST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });

      it('should apply DESC NULLS_LAST sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          { sorting: [{ field: 'numberType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_LAST }] },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });

      it('should apply multiple sorts', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectUpdateSQLSnapshot(
          {
            sorting: [
              { field: 'numberType', direction: SortDirection.ASC },
              { field: 'boolType', direction: SortDirection.DESC },
              { field: 'stringType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_FIRST },
              { field: 'dateType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_LAST },
            ],
          },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });
    });
  });

  describe('#delete', () => {
    const expectDeleteSQLSnapshot = (query: Query<TestEntity>, whereBuilder: WhereBuilder<TestEntity>): void => {
      const selectQueryBuilder = getEntityQueryBuilder(TestEntity, whereBuilder).delete(query);
      expectSQLSnapshot(selectQueryBuilder);
    };

    describe('with filter', () => {
      it('should call whereBuilder#build if there is a filter', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        const query = { filter: { stringType: { eq: 'foo' } } };
        when(mockWhereBuilder.build(anything(), query.filter, deepEqual({}), undefined)).thenCall(
          (where: WhereExpression) => where.andWhere(`stringType = 'foo'`),
        );
        expectDeleteSQLSnapshot(query, instance(mockWhereBuilder));
      });
    });
    describe('with paging', () => {
      it('should ignore paging args', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectDeleteSQLSnapshot({ paging: { limit: 10, offset: 11 } }, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });
    });

    describe('with sorting', () => {
      it('should ignore sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestEntity>>(WhereBuilder);
        expectDeleteSQLSnapshot(
          {
            sorting: [
              { field: 'numberType', direction: SortDirection.ASC },
              { field: 'boolType', direction: SortDirection.DESC },
              { field: 'stringType', direction: SortDirection.ASC, nulls: SortNulls.NULLS_FIRST },
              { field: 'dateType', direction: SortDirection.DESC, nulls: SortNulls.NULLS_LAST },
            ],
          },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });
    });
  });

  describe('#softDelete', () => {
    const expectSoftDeleteSQLSnapshot = (
      query: Query<TestSoftDeleteEntity>,
      whereBuilder: WhereBuilder<TestSoftDeleteEntity>,
    ): void => {
      const selectQueryBuilder = getEntityQueryBuilder(TestSoftDeleteEntity, whereBuilder).softDelete(query);
      expectSQLSnapshot(selectQueryBuilder);
    };

    describe('with filter', () => {
      it('should call whereBuilder#build if there is a filter', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestSoftDeleteEntity>>(WhereBuilder);
        const query = { filter: { stringType: { eq: 'foo' } } };
        when(mockWhereBuilder.build(anything(), query.filter, deepEqual({}), undefined)).thenCall(
          (where: WhereExpression) => where.andWhere(`stringType = 'foo'`),
        );
        expectSoftDeleteSQLSnapshot(query, instance(mockWhereBuilder));
      });
    });
    describe('with paging', () => {
      it('should ignore paging args', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestSoftDeleteEntity>>(WhereBuilder);
        expectSoftDeleteSQLSnapshot({ paging: { limit: 10, offset: 11 } }, instance(mockWhereBuilder));
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });
    });

    describe('with sorting', () => {
      it('should ignore sorting', () => {
        const mockWhereBuilder = mock<WhereBuilder<TestSoftDeleteEntity>>(WhereBuilder);
        expectSoftDeleteSQLSnapshot(
          {
            sorting: [
              { field: 'stringType', direction: SortDirection.ASC },
              { field: 'testEntityPk', direction: SortDirection.DESC },
            ],
          },
          instance(mockWhereBuilder),
        );
        verify(mockWhereBuilder.build(anything(), anything(), anything())).never();
      });
    });
  });
});
