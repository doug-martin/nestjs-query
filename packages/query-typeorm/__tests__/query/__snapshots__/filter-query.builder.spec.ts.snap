// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterQueryBuilder #delete with filter should call whereBuilder#build if there is a filter 1`] = `DELETE FROM "test_entity" WHERE "string_type" = 'foo'`;

exports[`FilterQueryBuilder #delete with filter should call whereBuilder#build if there is a filter 2`] = `Array []`;

exports[`FilterQueryBuilder #delete with paging should ignore paging args 1`] = `DELETE FROM "test_entity"`;

exports[`FilterQueryBuilder #delete with paging should ignore paging args 2`] = `Array []`;

exports[`FilterQueryBuilder #delete with sorting should ignore sorting 1`] = `DELETE FROM "test_entity"`;

exports[`FilterQueryBuilder #delete with sorting should ignore sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with filter should call whereBuilder#build if there is a filter 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" WHERE "TestEntity"."string_type" = 'foo'`;

exports[`FilterQueryBuilder #select with filter should call whereBuilder#build if there is a filter 2`] = `Array []`;

exports[`FilterQueryBuilder #select with filter should not call whereBuilder#build 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity"`;

exports[`FilterQueryBuilder #select with filter should not call whereBuilder#build 2`] = `Array []`;

exports[`FilterQueryBuilder #select with paging should apply empty paging args 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity"`;

exports[`FilterQueryBuilder #select with paging should apply empty paging args 2`] = `Array []`;

exports[`FilterQueryBuilder #select with paging should apply paging args going backward 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" LIMIT 10 OFFSET 10`;

exports[`FilterQueryBuilder #select with paging should apply paging args going backward 2`] = `Array []`;

exports[`FilterQueryBuilder #select with paging should apply paging args going forward 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" LIMIT 10 OFFSET 11`;

exports[`FilterQueryBuilder #select with paging should apply paging args going forward 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply ASC NULLS_FIRST sorting 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" ASC NULLS FIRST`;

exports[`FilterQueryBuilder #select with sorting should apply ASC NULLS_FIRST sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply ASC NULLS_LAST sorting 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" ASC NULLS LAST`;

exports[`FilterQueryBuilder #select with sorting should apply ASC NULLS_LAST sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply ASC sorting 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" ASC`;

exports[`FilterQueryBuilder #select with sorting should apply ASC sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply DESC NULLS_FIRST sorting 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" DESC NULLS FIRST`;

exports[`FilterQueryBuilder #select with sorting should apply DESC NULLS_FIRST sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply DESC NULLS_LAST sorting 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" DESC NULLS LAST`;

exports[`FilterQueryBuilder #select with sorting should apply DESC NULLS_LAST sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply DESC sorting 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" DESC`;

exports[`FilterQueryBuilder #select with sorting should apply DESC sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #select with sorting should apply multiple sorts 1`] = `SELECT "TestEntity"."test_entity_pk" AS "TestEntity_test_entity_pk", "TestEntity"."string_type" AS "TestEntity_string_type", "TestEntity"."bool_type" AS "TestEntity_bool_type", "TestEntity"."number_type" AS "TestEntity_number_type", "TestEntity"."date_type" AS "TestEntity_date_type", "TestEntity"."oneTestRelationTestRelationPk" AS "TestEntity_oneTestRelationTestRelationPk" FROM "test_entity" "TestEntity" ORDER BY "TestEntity"."number_type" ASC, "TestEntity"."bool_type" DESC, "TestEntity"."string_type" ASC NULLS FIRST, "TestEntity"."date_type" DESC NULLS LAST`;

exports[`FilterQueryBuilder #select with sorting should apply multiple sorts 2`] = `Array []`;

exports[`FilterQueryBuilder #softDelete with filter should call whereBuilder#build if there is a filter 1`] = `UPDATE "test_soft_delete_entity" SET "deleted_at" = CURRENT_TIMESTAMP WHERE "string_type" = 'foo'`;

exports[`FilterQueryBuilder #softDelete with filter should call whereBuilder#build if there is a filter 2`] = `Array []`;

exports[`FilterQueryBuilder #softDelete with paging should ignore paging args 1`] = `UPDATE "test_soft_delete_entity" SET "deleted_at" = CURRENT_TIMESTAMP`;

exports[`FilterQueryBuilder #softDelete with paging should ignore paging args 2`] = `Array []`;

exports[`FilterQueryBuilder #softDelete with sorting should ignore sorting 1`] = `UPDATE "test_soft_delete_entity" SET "deleted_at" = CURRENT_TIMESTAMP`;

exports[`FilterQueryBuilder #softDelete with sorting should ignore sorting 2`] = `Array []`;

exports[`FilterQueryBuilder #update with filter should call whereBuilder#build if there is a filter 1`] = `UPDATE "test_entity" SET "string_type" = ? WHERE "string_type" = 'foo'`;

exports[`FilterQueryBuilder #update with filter should call whereBuilder#build if there is a filter 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with paging should ignore paging args 1`] = `UPDATE "test_entity" SET "string_type" = ?`;

exports[`FilterQueryBuilder #update with paging should ignore paging args 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply ASC NULLS_FIRST sorting 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" ASC NULLS FIRST`;

exports[`FilterQueryBuilder #update with sorting should apply ASC NULLS_FIRST sorting 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply ASC NULLS_LAST sorting 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" ASC NULLS LAST`;

exports[`FilterQueryBuilder #update with sorting should apply ASC NULLS_LAST sorting 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply ASC sorting 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" ASC`;

exports[`FilterQueryBuilder #update with sorting should apply ASC sorting 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply DESC NULLS_FIRST sorting 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" DESC NULLS FIRST`;

exports[`FilterQueryBuilder #update with sorting should apply DESC NULLS_FIRST sorting 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply DESC NULLS_LAST sorting 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" DESC NULLS LAST`;

exports[`FilterQueryBuilder #update with sorting should apply DESC NULLS_LAST sorting 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply DESC sorting 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" DESC`;

exports[`FilterQueryBuilder #update with sorting should apply DESC sorting 2`] = `
Array [
  baz,
]
`;

exports[`FilterQueryBuilder #update with sorting should apply multiple sorts 1`] = `UPDATE "test_entity" SET "string_type" = ? ORDER BY "number_type" ASC, "bool_type" DESC, "string_type" ASC NULLS FIRST, "date_type" DESC NULLS LAST`;

exports[`FilterQueryBuilder #update with sorting should apply multiple sorts 2`] = `
Array [
  baz,
]
`;
